docs/concurrency.org

```diff
@@ -1,68 +1,251 @@
+  printf("Everything is done.\n");
+
+  return 0;
+}
+#+end_src
+
+Or you could write your own wrapper kind of like the way Zig does (this is not going to work on Windows, but you get the idea):
+#+begin_src c
+#include <stdio.h>
+
+#ifdef _WIN32
+#include <windows.h>
+#else
+#include <pthread.h>
+#endif
+
+int myCreate(unsigned long *thread, void *func) {
+#ifdef _WIN32
+  return hThread = CreateThread(NULL, 0, func, NULL, 0, thread);
+#else
+  return pthread_create(thread, NULL, func, NULL);
+#endif
+}
+
+void myJoin(unsigned long thread) {
+#ifdef _WIN32
+  return WaitForSingleObject(thread, INFINITE);
+#else
+  pthread_join(thread, NULL);
+#endif
+}
+
+void *ThreadFunc(void *arg) {
+  printf("Thread running...\n");
+  return NULL;
+}
+
+int main() {
+  pthread_t thread; // TODO I should also wrap that which is pthread specific
+
+  int rc = myCreate(&thread, ThreadFunc);
+  if (rc) {
+    printf("Failed to create thread. Return code: %d\n", rc);
+    return 1;
+  }
+
+  myJoin(thread);
+
+  printf("Everything is done.\n");
+
+  return 0;
+}
+#+end_src
+
+**** Thread synchronization
+
+
+**** Leaky abstraction
+TODO: est ce que je peux fine tuned threads en Zig ? (priority, stack size, ...)
+
+**** Conclusion
+Zig threads are really useful since they have a very user-friendly abstraction with not a lot of functionalites that are almost never used anyway. This abstraction is also very useful for what we saw earlier, you don't have to worry about the target system, Zig is going to choose the right implementation for you.
+
+But this leaky abstraction comes at a cost, you can not fine-tune your threads as much as you would like to.
+
+If you need specific thread functionalities, like the ones we talked about, you can still do that in Zig by wrapping the C pthread library for exemple or directly use the OS native threads you want.
+
+*** async/await
+This method uses suspensible stackless coroutines, this solution does not necessarly mean that you are going to have multiple threads or parallelism.
+
+We are not going to dive deeper into this solution because it has been deprecated since 0.11 and is not coming back soon.
+
+However it is still a good reading and way to understand concurency to read this [[https://zig.guide/async/introduction/][very good guide]] that was made for this solution.
+By reading this you might notice that **async/await** might **never** come out.
+
+Note that if this solution is to be brought again it might come with breaking changes, so the syntax might change.
+
+You can find a [[https://github.com/ziglang/zig/issues/6025][Github discussion]] about the progress of this feature and why it is not implemented in the current version.
+
+You can see [[https://ziglang.org/download/0.12.0/release-notes.html#AsyncAwait-Feature-Status][here]] the main reasons why this solution is not implemented yet.
```

Comment:
Add citation in citeproc

Reply:
https://github.com/Pismice/HEIG_ZIG/pull/11#discussion_r1607147705